### PR TITLE
PWGHF: Prepare analysis of omega_c with strangeness tracking

### PR DESCRIPTION
**Pull request by qgp**

### CODE REVIEW

File: PWGHF/D2H/Tasks/taskOmegacSt.cxx [hunk 1]
```diff
@@ -18,6 +18,10 @@
 #include <TPDGCode.h>
 
 #include "CCDB/BasicCCDBManager.h"
+#include "Common/Core/RecoDecay.h"
+#include "Common/Core/trackUtilities.h"
+#include "Common/DataModel/PIDResponse.h"
+#include "Common/DataModel/TrackSelectionTables.h"
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskOmegacSt.cxx:24]: These should be kept after O2 includes.

File: PWGHF/D2H/Tasks/taskOmegacSt.cxx [hunk 2]
```diff
@@ -29,20 +33,116 @@
...
 #include "PWGHF/Core/PDG.h"
 #include "PWGHF/Core/SelectorCuts.h"
 
 using namespace o2;
 using namespace o2::framework;
 using namespace o2::framework::expressions;
 
+namespace o2::aod
+{
+namespace stomegac
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskOmegacSt.cxx:46]:
```suggestion
namespace st_omegac
```

Comment (vkucera) [inline PWGHF/D2H/Tasks/taskOmegacSt.cxx:46]: Why is the data model defined directly in the task file?

File: PWGHF/D2H/Tasks/taskOmegacSt.cxx [hunk 3]
```diff
@@ -29,20 +33,116 @@
...
+{
+DECLARE_SOA_COLUMN(MassOmegaC, massOmegaC, float);
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskOmegacSt.cxx:48]:
```suggestion
DECLARE_SOA_COLUMN(MassOmegac, massOmegac, float);
```

File: PWGHF/D2H/Tasks/taskOmegacSt.cxx [hunk 4]
```diff
@@ -29,20 +33,116 @@
...
+DECLARE_SOA_COLUMN(DcaZPi, dcaZPi, float);
+DECLARE_SOA_COLUMN(Chi2TopologicalOmegaC, chi2TopologicalOmegaC, float);
+DECLARE_SOA_COLUMN(Chi2TopologicalOmega, chi2TopologicalOmega, float);
+DECLARE_SOA_COLUMN(DecayLengthOmegaC, decayLengthOmegaC, float);
+DECLARE_SOA_COLUMN(DecayLengthXYOmegaC, decayLengthXYOmegaC, float);
+DECLARE_SOA_COLUMN(DecayLengthOmega, decayLengthOmega, float);
+DECLARE_SOA_COLUMN(DecayLengthXYOmega, decayLengthXYOmega, float);
+} // namespace stomegac
+
+namespace stomegacgen
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskOmegacSt.cxx:84]:
```suggestion
namespace st_omegac_gen
```

File: PWGHF/D2H/Tasks/taskOmegacSt.cxx [hunk 5]
```diff
@@ -29,20 +33,116 @@
...
+                  stomegac::DcaZKa,
+                  stomegac::DcaXYPi,
+                  stomegac::DcaZPi,
+                  stomegac::Chi2TopologicalOmegaC,
+                  stomegac::Chi2TopologicalOmega,
+                  stomegac::DecayLengthOmegaC,
+                  stomegac::DecayLengthXYOmegaC,
+                  stomegac::DecayLengthOmega,
+                  stomegac::DecayLengthXYOmega);
+DECLARE_SOA_TABLE(HfOmegaStGen, "AOD", "HFOMEGACSTGEN",
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskOmegacSt.cxx:133]:
```suggestion
                  stomegac::DecayLengthXYOmega);

DECLARE_SOA_TABLE(HfOmegaStGen, "AOD", "HFOMEGACSTGEN",
```

File: PWGHF/D2H/Tasks/taskOmegacSt.cxx [hunk 6]
```diff
@@ -129,14 +236,9 @@ struct HfTaskOmegacSt {
...
-               aod::McCollisions const&,
-               aod::AssignedTrackedCascades const& trackedCascades,
-               aod::Cascades const&,
-               aod::V0s const&,
-               TracksExt const& tracks,
-               aod::McParticles const&,
-               aod::BCsWithTimestamps const&)
+  void processData(aod::Collision const& collision,
+                   aod::AssignedTrackedCascades const& trackedCascades, aod::Cascades const& cascades,
+                   aod::V0s const& v0s, TracksExt const& tracks, aod::BCsWithTimestamps const&)
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskOmegacSt.cxx:241]: Keep one argument per line for better readability.

File: PWGHF/D2H/Tasks/taskOmegacSt.cxx [hunk 7]
```diff
@@ -164,20 +266,165 @@ struct HfTaskOmegacSt {
...
+      std::array<double, 3> m{o2::analysis::pdg::MassProton, o2::analysis::pdg::MassPiMinus, o2::analysis::pdg::MassKMinus};
+      std::array<std::array<float, 3>, 3> p;
+      // TODO: treat properly as omega +-
+      p[0] = {v0TrackPos.px(), v0TrackPos.py(), v0TrackPos.pz()};
+      p[1] = {v0TrackNeg.px(), v0TrackNeg.py(), v0TrackNeg.pz()};
+      p[2] = {bachelor.px(), bachelor.py(), bachelor.pz()};
+      double massOmega1 = RecoDecay::m(p, m);
+      p[0] = {v0TrackNeg.px(), v0TrackNeg.py(), v0TrackNeg.pz()};
+      p[1] = {v0TrackPos.px(), v0TrackPos.py(), v0TrackPos.pz()};
+      p[2] = {bachelor.px(), bachelor.py(), bachelor.pz()};
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskOmegacSt.cxx:284]: Not needed, the bachelor does not change.

File: PWGHF/D2H/Tasks/taskOmegacSt.cxx [hunk 8]
```diff
@@ -164,20 +266,165 @@ struct HfTaskOmegacSt {
...
+
+      std::array<double, 3> m{o2::analysis::pdg::MassProton, o2::analysis::pdg::MassPiMinus, o2::analysis::pdg::MassKMinus};
+      std::array<std::array<float, 3>, 3> p;
+      // TODO: treat properly as omega +-
+      p[0] = {v0TrackPos.px(), v0TrackPos.py(), v0TrackPos.pz()};
+      p[1] = {v0TrackNeg.px(), v0TrackNeg.py(), v0TrackNeg.pz()};
+      p[2] = {bachelor.px(), bachelor.py(), bachelor.pz()};
+      double massOmega1 = RecoDecay::m(p, m);
+      p[0] = {v0TrackNeg.px(), v0TrackNeg.py(), v0TrackNeg.pz()};
+      p[1] = {v0TrackPos.px(), v0TrackPos.py(), v0TrackPos.pz()};
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskOmegacSt.cxx:283]: Would `std::swap` work with this?

File: PWGHF/D2H/Tasks/taskOmegacSt.cxx [hunk 9]
```diff
@@ -164,20 +266,165 @@ struct HfTaskOmegacSt {
...
+          }
+        }
+      }
+    }
+  }
+  PROCESS_SWITCH(HfTaskOmegacSt, processData, "Process data", true);
+
+  void processGen(aod::Collision const& collision, aod::McCollisions const& mcCollisions,
+                  aod::AssignedTrackedCascades const& trackedCascades, aod::Cascades const& cascades,
+                  aod::V0s const& v0s, TracksExtMc const& tracks, aod::McParticles const& mcParticles, aod::BCsWithTimestamps const&)
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskOmegacSt.cxx:395]:
```suggestion
  void processMcGen(aod::Collision const& collision, aod::McCollisions const& mcCollisions,
                  aod::AssignedTrackedCascades const& trackedCascades, aod::Cascades const& cascades,
                  aod::V0s const& v0s, TracksExtMc const& tracks, aod::McParticles const& mcParticles, aod::BCsWithTimestamps const&)
```
Keep one argument per line.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/D2H/Tasks/taskOmegacSt.cxx (hunk 2 above)
Comment (qgp) [reply PWGHF/D2H/Tasks/taskOmegacSt.cxx:46]: This is the definition of the tree output of that specific task and I followed the example of other tasks having this in the same source file. Are the definitions for tree outputs normally added under `DataModel`?

Comment (qgp) [reply PWGHF/D2H/Tasks/taskOmegacSt.cxx:46]: Ok, so we should move the workflow to `TableProducer`. Do we also have to split the table definition into a separate file under `DataModel`?

Comment (vkucera) [reply PWGHF/D2H/Tasks/taskOmegacSt.cxx:46]: If the tables are (supposed to be) consumed by another workflow, it has to be defined in `DataModel`, if not then it is a tree creator workflow and the tables can stay inside the file.

Comment (qgp) [reply PWGHF/D2H/Tasks/taskOmegacSt.cxx:46]: Ok, it falls in the category of tree creator. What should the workflow be called then? data-creator-st-omegac?

Comment (vkucera) [reply PWGHF/D2H/Tasks/taskOmegacSt.cxx:46]: `tree-creator-omegac-st` and the file should go to `PWGHF/TableProducer`.
(`data-creator` workflows have been introduced with the reduced data format for B mesons, which are a different kind of table producer for self-contained tables consumed by analysis tasks.)

**Comment by qgp** [comment]
Thanks @vkucera, updated with your suggestions.

**Comment by vkucera** [comment]
Thanks @qgp for implementing the changes. I would still be in favour of moving the file in `PWGHF/TableProducer` but I let @fcatalan92 @mfaggin decide.

**Comment by qgp** [comment]
@vkucera You mean `PWGHF/TableProducer` instead of `PWGHF/D2H/TableProducer`?

**Comment by vkucera** [comment]
> @vkucera You mean `PWGHF/TableProducer` instead of `PWGHF/D2H/TableProducer`?

Yes, since all the Run 3 tree creators are there.

**Comment by qgp** [comment]
Ok, then I had misunderstood. Of course, I can move it there, just let me know what the preference is.

**Comment by qgp** [comment]
Moved to `PWGHF/TableProducer` now.

**Review by vkucera — APPROVED** [review]

**Review by fcatalan92 — APPROVED** [review]
Hi @qgp, all looks good to me!
At a later stage, I think you should split the workflow in the usual parts (skimming, candidate creation, candidate selection, ...). This would allow us to have part of the workflow in common with the analysis without strangeness tracking, which is planned.

**Comment by qgp** [comment]
Hi @fcatalan92, thanks! I think the splitting (and exploiting of commonalities) makes sense once we have verified the basics for strangeness tracking.

**Pull request merged**